### PR TITLE
MobileNetv2 

### PR DESCRIPTION
**Pull request by Arjun31415**
Work in progress. Model structure seems to be correct, could be simplified more by getting rid of a few enums. 

Need help in loading pytorch weight 🥲

### CODE REVIEW

File: mobilenet-burn/src/model/inverted_residual.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            ),
+            InvertedResidualSequentialType::Conv(
+                Conv2dConfig::new([hidden_dim, self.oup], [1, 1])
+                    .with_stride([1, 1])
+                    .with_padding(burn::nn::PaddingConfig2d::Explicit(0, 0))
+                    .with_bias(false)
+                    .init(device),
+            ),
+            match self.norm_type {
+                crate::model::conv_norm::NormalizationType::BatchNorm(_) => {
```
Comment (laggui) [inline mobilenet-burn/src/model/inverted_residual.rs:59]: `NormalizationType` is already in scope, we can remove the `crate::model::conv_norm::`

File: mobilenet-burn/src/model/inverted_residual.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    NormLayer(NormalizationLayer<B, 4>),
+}
+
+#[derive(Module, Debug)]
+pub struct InvertedResidual<B: Backend> {
+    use_res_connect: bool,
+    layers: Vec<InvertedResidualSequentialType<B>>,
+    out_channels: usize,
+    is_cn: bool,
+}
```
Comment (laggui) [inline mobilenet-burn/src/model/inverted_residual.rs:24]: I'd add a small docstring, could simple point to a reference (e.g, https://paperswithcode.com/method/inverted-residual-block)

File: mobilenet-burn/src/model/inverted_residual.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    ConvNormActivation(Conv2dNormActivation<B>),
+    NormLayer(NormalizationLayer<B, 4>),
+}
+
+#[derive(Module, Debug)]
+pub struct InvertedResidual<B: Backend> {
+    use_res_connect: bool,
+    layers: Vec<InvertedResidualSequentialType<B>>,
+    out_channels: usize,
+    is_cn: bool,
```
Comment (laggui) [inline mobilenet-burn/src/model/inverted_residual.rs:23]: `is_cn` and `out_channels` are never actually used it seems?

File: mobilenet-burn/src/model/inverted_residual.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+use burn::nn::conv::Conv2dConfig;
+use burn::nn::BatchNormConfig;
+use burn::tensor::Tensor;
+use burn::{module::Module, nn::conv::Conv2d, tensor::backend::Backend};
+#[derive(Module, Debug)]
+enum InvertedResidualSequentialType<B: Backend> {
+    Conv(Conv2d<B>),
+    ConvNormActivation(Conv2dNormActivation<B>),
+    NormLayer(NormalizationLayer<B, 4>),
+}
```
Comment (laggui) [inline mobilenet-burn/src/model/inverted_residual.rs:16]: I know that the pytorch implementation uses `nn.Sequential`, but the number of layers isn't even variable. The only thing that changes is an additional `Conv2dNormActivation` when expand ratio != 1. I think sequential is overused in general 😄 

In this case I would probably remove the enum and layers vector to have a fixed definition. For example:

```rust
#[derive(Module, Debug)]
pub struct PointWiseLinear<B: Backend> {
   conv: Conv2d<B>,
   norm: NormalizationLayer<B, 4>,
}

#[derive(Module, Debug)]
pub struct InvertedResidual<B: Backend> {
    use_res_connect: bool,
    pw: Option<Conv2dNormActivation<B>>, // pointwise, only when expand ratio != 1
    dw: Conv2dNormActivation<B>,
    pw_linear: PointWiseLinear<B>,
}
```

And for the weights we'll be able to remap them anyway.

File: mobilenet-burn/src/model/inverted_residual.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            match layer {
+                InvertedResidualSequentialType::Conv(conv) => out = conv.forward(out),
+                InvertedResidualSequentialType::ConvNormActivation(conv_norm) => {
+                    out = conv_norm.forward(out)
+                }
+                InvertedResidualSequentialType::NormLayer(NormalizationLayer::BatchNorm(x)) => {
+                    out = x.forward(out)
+                }
+            }
+        }
```
Comment (laggui) [inline mobilenet-burn/src/model/inverted_residual.rs:88]: If we do apply the suggested change above (no more vec of enum to simulate a sequential type), then this will be simplified.

File: mobilenet-burn/src/model/mobilenet.rs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+#[derive(Module, Debug)]
+enum FeatureLayersType<B: Backend> {
+    Block(InvertedResidual<B>),
+    Conv(Conv2dNormActivation<B>),
+}
+#[derive(Module, Debug)]
+enum ClassifierLayersType<B: Backend> {
+    Dropout(Dropout),
+    Linear(Linear<B>),
+}
```
Comment (laggui) [inline mobilenet-burn/src/model/mobilenet.rs:58]: (Not specific to this block only)

We can't specify a minimum number of lines between elements with rustfmt (only in the nightly), but we try to be consistent across the code 🙂 one line break across should be good

File: mobilenet-burn/src/model/mobilenet.rs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                vec![6, 160, 3, 2],
+                vec![6, 320, 1, 1],
+            ]
+        }
+        if inverted_residual_setting[0].len() != 4 {
+            panic!(
+                "inverted_residual_setting should be non-empty or a 4-element list, got {:#?}",
+                self.inverted_residual_setting
+            )
+        }
```
Comment (laggui) [inline mobilenet-burn/src/model/mobilenet.rs:81]: This can probably just be a const array definition.

File: mobilenet-burn/src/model/utils.rs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+use core::cmp::max;
+
+pub fn make_divisble(v: f32, divisor: i32, min_value: Option<i32>) -> i32 {
+    let min_value = match min_value {
+        Some(min_value) => min_value,
+        None => divisor,
+    };
```
Comment (laggui) [inline mobilenet-burn/src/model/utils.rs:7]: Seems like `min_value` is always none, so we could simplify.

File: mobilenet-burn/src/model/mobilenet.rs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        let mut input_channel = make_divisble(
+            input_channel as f32 * self.width_mult,
+            self.round_nearest as i32,
```
Comment (laggui) [inline mobilenet-burn/src/model/mobilenet.rs:84]: Should `round_nearest` really be configurable? Or just keep it as a const with value of `8` as with the standard architecture variants?

File: mobilenet-burn/src/model/mobilenet.rs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+                .with_kernel_size(1)
+                .init(device),
+        ));
+
+        let classifier = vec![
+            ClassifierLayersType::Dropout(DropoutConfig::new(self.dropout).init()),
+            ClassifierLayersType::Linear(
+                LinearConfig::new(last_channel, self.num_classes).init(device),
+            ),
+        ];
```
Comment (laggui) [inline mobilenet-burn/src/model/mobilenet.rs:132]: Instead of using a vec here for a static definition that always contains a linear and dropout, we should add these modules directly to the model fields.

File: mobilenet-burn/src/model/mobilenet.rs [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    norm_layer: NormalizationType,
+
+    #[config(default = "0.2")]
+    dropout: f64,
+}
+#[derive(Module, Debug)]
+enum FeatureLayersType<B: Backend> {
+    Block(InvertedResidual<B>),
+    Conv(Conv2dNormActivation<B>),
+}
```
Comment (laggui) [inline mobilenet-burn/src/model/mobilenet.rs:53]: Feature layers are convolutional blocks, so we could rename the enum to something `ConvBlock` to be a bit more explicit.

```rust
enum ConvBlock<B: Backend> {
    InvertedResidual(InvertedResidual<B>),
    Conv(Conv2dNormActivation<B>),
}
```

**Review by laggui — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution!

Overall, the implementation looks good. I have a couple of comments, mainly related to the fact that the pytorch implementation was followed a bit too closely 😄 

If you have any questions for loading the pre-trained weights feel free to ping me on discord! Otherwise I can take care of that afterwards, no worries.

File: mobilenet-burn/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+[features]
+default = []
+std = []
+pretrained = ["burn/network", "std", "dep:dirs"]
+
+[dependencies]
+# Note: default-features = false is needed to disable std
+burn = { git = "https://github.com/tracel-ai/burn.git", rev = "c729401fb268ae0451ed2695f1b31653fb9207de", default-features = false }
+burn-import = { git = "https://github.com/tracel-ai/burn.git", rev = "c729401fb268ae0451ed2695f1b31653fb9207de" }
```
[the remark this answers is not in the document]
Comment (Arjun31415) [reply mobilenet-burn/Cargo.toml:16]: this would require a rewrite right ?. Should I do this after I resolve the other issues ?

File: mobilenet-burn/src/model/mobilenet.rs [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+//         weights: &weights::Weights,
+//         device: &Device<B>,
+//     ) -> Result<MobileNetV2Record<B>, RecorderError> {
+//         // Download torch weights
+//         let torch_weights = weights.download().map_err(|err| {
+//             RecorderError::Unknown(format!("Could not download weights.\nError: {err}"))
+//         })?;
+//
+//         // Load weights from torch state_dict
+//         let load_args = LoadArgs::new(torch_weights)
```
Comment (Arjun31415) [inline mobilenet-burn/src/model/mobilenet.rs:224]: this was just copied from resnet XD

Comment (laggui) [reply mobilenet-burn/src/model/mobilenet.rs:196]: I can take over after the other comments are addressed to add the weights loading. It'll probably follow a very similar pattern to the resnet, the most important part will be remapping the keys properly 🙂

File: mobilenet-burn/src/model/inverted_residual.rs (hunk 3 above)
Comment (Arjun31415) [reply mobilenet-burn/src/model/inverted_residual.rs:23]: yeah, because I was not sure how to use it and reading the pytorch code only made me more confused. I think removing it is better. (Ref:https://github.com/pytorch/vision/blob/main/torchvision/models/mobilenetv2.py)

File: mobilenet-burn/src/model/mobilenet.rs (hunk 7 above)
Comment (Arjun31415) [reply mobilenet-burn/src/model/mobilenet.rs:93]: Yes that can be done. It would also eliminate the need for the `panic!` in the next few lines

File: mobilenet-burn/src/model/mobilenet.rs (hunk 9 above)
Comment (Arjun31415) [reply mobilenet-burn/src/model/mobilenet.rs:90]: I'm not sure, I too thought it would be better to have it as a constant 8

File: mobilenet-burn/src/model/inverted_residual.rs (hunk 4 above)
Comment (Arjun31415) [reply mobilenet-burn/src/model/inverted_residual.rs]: Done

File: mobilenet-burn/src/model/mobilenet.rs (hunk 10 above)
Comment (Arjun31415) [reply mobilenet-burn/src/model/mobilenet.rs]: Done. I made it into another Classifier Struct

**Comment by Arjun31415** [comment]
Only the repo's README needs to be update

**Review by nathanielsimard — COMMENTED** [review]
Really good job, only a few minor comments.

File: mobilenetv2-burn/src/model/imagenet.rs [hunk 14]
```diff
@@ -0,0 +1,1034 @@
+use burn::tensor::{backend::Backend, Device, Tensor};
```
[the remark this answers is not in the document]
Comment (Arjun31415) [reply mobilenetv2-burn/src/model/imagenet.rs:1]: It is used in resnet and now in mobilenet. I too feel that we could classify the models say into image classification module and then have common utils, then have the models themselves be sub modules in it. It would also be helpful in future.

[the remark this answers is not in the document]
Comment (laggui) [reply mobilenetv2-burn/src/model/imagenet.rs:1]: Yeah it's used in both ResNet and MobileNet.. but right now the models are released separately. We should probably finish the discussion in #16 for loading weights since this is where the common code lies.

Not sure what the best approach would be.

**Comment by laggui** [comment]
@Arjun31415 just merged 🎉 Thanks for your contribution!

**Comment by Arjun31415** [comment]
time for mobilenetv3 :D
